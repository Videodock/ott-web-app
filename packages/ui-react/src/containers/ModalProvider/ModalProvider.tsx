import { createContext, type PropsWithChildren, type RefObject, useCallback, useEffect, useMemo, useState } from 'react';
import useEventCallback from '@jwp/ott-hooks-react/src/useEventCallback';

import scrollbarSize from '../../utils/dom';

type Modal = {
  modalId: string;
  modalRef: RefObject<HTMLElement>;
  onClose?: () => void;
  containerRef?: RefObject<HTMLElement>;
  lastFocus: Element | null;
};

type ModalContextValue = {
  openModal: (modalId: string, modalRef: RefObject<HTMLElement>, onClose?: () => void, containerRef?: RefObject<HTMLElement>) => void;
  closeModal: (modalId: string, notify?: boolean) => void;
  closeAllModals: (notify?: boolean) => void;
  modals: Modal[];
};

export const ModalContext = createContext<ModalContextValue>({
  openModal() {
    throw new Error('Not implemented');
  },
  closeModal() {
    throw new Error('Not implemented');
  },
  closeAllModals() {
    throw new Error('Not implemented');
  },
  modals: [],
});

const byId = (modalId: string) => (modal: Modal) => modal.modalId === modalId;
const notById = (modalId: string) => (modal: Modal) => modal.modalId !== modalId;

const focusModal = (openedModal: Modal, targetElement?: Element | null) => {
  const { modalRef } = openedModal;

  requestAnimationFrame(() => {
    if (!modalRef.current) return;

    // prefer focusing the targetElement
    if (targetElement && targetElement instanceof HTMLElement && modalRef.current.contains(targetElement)) {
      return targetElement.focus();
    }

    // find the first interactive element
    const interactiveElement = modalRef.current.querySelectorAll('input, a, button, [tabindex]')[0] as HTMLElement | null;

    interactiveElement?.focus();
  });
};

const elementIsFocusable = (element: Element | null): element is HTMLElement => {
  const inertElements = document.querySelectorAll('[inert]');
  const parentHasInert = Array.from(inertElements).some((parent) => parent.contains(element));

  return element instanceof HTMLElement && document.body.contains(element) && !parentHasInert;
};

const restoreFocus = (closedModal: Modal, activeModals: Modal[]) => {
  const activeModal = activeModals[activeModals.length - 1];

  // if there is still an active modal, focus that
  if (activeModal) {
    return focusModal(activeModal, closedModal.lastFocus);
  }

  // temp variable because originFocusElement gets cleared
  const originFocus = originFocusElement;

  // focus the last focussed or origin element with fallback to the body element
  requestAnimationFrame(() => {
    if (elementIsFocusable(closedModal.lastFocus)) {
      closedModal.lastFocus.focus({ preventScroll: true });
    } else if (elementIsFocusable(originFocus)) {
      originFocus.focus({ preventScroll: true });
    } else if (document.activeElement instanceof HTMLElement) {
      document.activeElement.blur();
      window.focus();
    }
  });
};

let originFocusElement: HTMLElement | null = null;
let modalInertElements: HTMLElement[] = [];

const handleInert = (activeModals: Modal[]) => {
  const appView = document.querySelector('#root') as HTMLDivElement | null;

  modalInertElements.forEach((element) => {
    element.inert = false;
  });

  modalInertElements = [];

  if (appView) {
    appView.inert = activeModals.some((modal) => !modal.containerRef);
  }

  activeModals.forEach((modal) => {
    if (modal.containerRef?.current) {
      modalInertElements.push(modal.containerRef.current);
      modal.containerRef.current.inert = true;
    }
  });
};

const handleBodyScrolling = (activeModals: Modal[]) => {
  if (activeModals.length > 0) {
    document.body.style.marginRight = `${scrollbarSize()}px`;
    document.body.style.overflowY = 'hidden';
  } else {
    document.body.style.removeProperty('margin-right');
    document.body.style.removeProperty('overflow-y');
  }
};

const ModalProvider = ({ children }: PropsWithChildren) => {
  const [modals, setModals] = useState<Modal[]>([]);

  const keyDownEventHandler = useEventCallback((event: globalThis.KeyboardEvent) => {
    if (event.key === 'Escape' && modals.length) {
      closeModal(modals[modals.length - 1].modalId);
    }
  });

  useEffect(() => {
    document.addEventListener('keydown', keyDownEventHandler);

    return () => {
      document.removeEventListener('keydown', keyDownEventHandler);
    };
  }, [keyDownEventHandler]);

  useEffect(() => {
    handleInert(modals);
    handleBodyScrolling(modals);

    // keep track of the origin focus element in case we stack multiple modals, for example, sidebar -> account modal
    if (modals.length === 1 && !originFocusElement) {
      originFocusElement = modals[0].lastFocus as HTMLElement;
    } else if (modals.length === 0) {
      originFocusElement = null;
    }
  }, [modals]);

  const openModal = useCallback((modalId: string, modalRef: RefObject<HTMLElement>, onClose?: () => void, containerRef?: RefObject<HTMLElement>) => {
    const modal: Modal = {
      modalId,
      onClose,
      modalRef,
      containerRef,
      lastFocus: document.activeElement,
    };

    focusModal(modal);
    setModals((current) => [...current, modal]);
  }, []);

  const closeModal = useCallback(
    (modalId: string, notify = true) => {
      const modal = modals.find(byId(modalId));
      const newModals = modals.filter(notById(modalId));

      if (modal) {
        if (notify) modal.onClose?.();
        restoreFocus(modal, newModals);
      }

      setModals(newModals);
    },
    [modals],
  );

  const closeAllModals = useCallback(
    (notify = true) => {
      // first modal opened has the lastFocus
      const modal = modals[0];

      if (notify) modals.forEach((modal) => modal.onClose?.());

      restoreFocus(modal, []);
      setModals([]);
    },
    [modals],
  );

  const value = useMemo(
    () => ({
      openModal,
      closeModal,
      closeAllModals,
      modals,
    }),
    [closeAllModals, closeModal, modals, openModal],
  );

  return <ModalContext.Provider value={value}>{children}</ModalContext.Provider>;
};

export default ModalProvider;
