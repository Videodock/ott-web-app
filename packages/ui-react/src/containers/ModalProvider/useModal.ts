import { type RefObject, useContext, useEffect, useRef } from 'react';
import useEventCallback from '@jwp/ott-hooks-react/src/useEventCallback';

import { ModalContext } from './ModalProvider';

type Params = {
  open: boolean;
  onClose?: () => void;
  modalRef: RefObject<HTMLElement>;
  containerRef?: RefObject<HTMLElement>;
};

export const useModal = ({ open, onClose, modalRef, containerRef }: Params) => {
  const modalId = useRef(String(Math.round(Math.random() * 1_000_000))).current;
  const { openModal, closeModal, modals } = useContext(ModalContext);
  const isOpen = modals.some((modal) => modal.modalId === modalId);

  const onCloseCallback = useEventCallback(onClose);

  useEffect(() => {
    if (open) {
      openModal(modalId, modalRef, onCloseCallback, containerRef);
    } else if (isOpen) {
      closeModal(modalId, false);
    }
    // only react to `open` changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [open]);

  useEffect(() => {
    return () => {
      isOpen && closeModal(modalId, false);
    };
    // unmount only to unregister the modal when it was open
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const close = () => closeModal(modalId);

  return {
    isOpen,
    closeModal: close,
    open,
  };
};
