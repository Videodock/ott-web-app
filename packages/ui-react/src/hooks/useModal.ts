import { useCallback, useEffect } from 'react';

import scrollbarSize from '../utils/dom';

const removeBodyScrolling = () => {
  document.body.style.marginRight = `${scrollbarSize(true)}px`;
  document.body.style.overflowY = 'hidden';
};

const restoreBodyScrolling = () => {
  const dialogs = Array.from(document.querySelectorAll('dialog'));
  const openDialogs = dialogs.filter((dialog) => dialog.open).length;

  if (!openDialogs) {
    document.body.style.removeProperty('margin-right');
    document.body.style.removeProperty('overflow-y');
  }
};

// shared variable between `useModal` usages
let originElement: HTMLElement | undefined;

export const useModal = () => {
  const restoreFocus = useCallback(() => {
    // restore focus when the focus is "lost" to the body element
    if (!originElement || document.activeElement !== document.body) {
      return;
    }

    const dialogs = Array.from(document.querySelectorAll('dialog'));
    const openDialogs = dialogs.filter((dialog) => dialog.open).length;

    // this was the last open dialog
    if (openDialogs === 0) {
      originElement.focus();
      originElement = undefined;
    }
  }, []);

  useEffect(() => {
    return () => {
      restoreBodyScrolling();
      restoreFocus();
    };
  }, [restoreFocus]);

  const handleOpen = useCallback(() => {
    if (!originElement) {
      originElement = document.activeElement as HTMLElement;
    }

    removeBodyScrolling();
  }, []);

  const handleClose = useCallback(() => {
    restoreFocus();
    restoreBodyScrolling();
  }, [restoreFocus]);

  return { handleOpen, handleClose };
};
