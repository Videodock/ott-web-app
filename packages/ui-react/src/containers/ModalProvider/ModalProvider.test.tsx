import React, { type RefObject, useRef } from 'react';
import { fireEvent, render } from '@testing-library/react';

import ModalProvider from './ModalProvider';
import { useModal } from './useModal';

type Props = {
  open: boolean;
  onClose?: () => void;
  modalRef: RefObject<HTMLElement>;
  containerRef?: RefObject<HTMLElement>;
};

const TestModal = ({ open, onClose, containerRef, name }: Omit<Props, 'modalRef'> & { name: string }) => {
  const modalRef = useRef<HTMLDivElement>(null);
  useModal({ open, onClose, containerRef, modalRef });

  if (!open) {
    return null;
  }

  return (
    <div data-testid={`modal-${name}`} ref={modalRef}>
      <button onClick={onClose}>close {name}</button>
    </div>
  );
};

const renderSingleModal = (open: boolean, onClose?: () => void) => {
  return (
    <ModalProvider>
      <>
        <div id="root" data-testid="root">
          <button onClick={vi.fn()}>last focus</button>
        </div>
        <TestModal open={open} onClose={onClose} name="modal-1" />
      </>
    </ModalProvider>
  );
};

const renderMultiModal = (open: boolean, secondOpen: boolean, onClose?: () => void, onSecondClose?: () => void) => {
  return (
    <ModalProvider>
      <>
        <div id="root" data-testid="root">
          <button onClick={vi.fn()}>last focus</button>
        </div>
        <TestModal open={open} onClose={onClose} name="modal-1" />
        <TestModal open={secondOpen} onClose={onSecondClose} name="modal-2" />
      </>
    </ModalProvider>
  );
};

describe('<ModalProvider />', () => {
  beforeAll(() => {
    vi.useFakeTimers();
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  test('renders provider with test modal', () => {
    const { container, getByTestId } = render(renderSingleModal(true));

    expect(container).toMatchSnapshot();
    expect(getByTestId('root')).toHaveAttribute('inert');
  });

  test('clears inert when all modals are closed', () => {
    const onClose = vi.fn();

    const { getByTestId, getByText, rerender } = render(renderSingleModal(true, onClose));

    expect(getByTestId('root')).toHaveAttribute('inert');

    fireEvent.click(getByText('close modal-1'));

    rerender(renderSingleModal(false, onClose));

    expect(onClose).toHaveBeenCalledTimes(1);
    expect(getByTestId('root')).not.toHaveAttribute('inert');
    expect(document.activeElement).toBe(document.body);
  });

  test('focus the first interactive element', () => {
    const onClose = vi.fn();

    const { getByText, rerender } = render(renderSingleModal(false, onClose));

    rerender(renderSingleModal(true, onClose));
    vi.runAllTimers();

    expect(document.activeElement).toBe(getByText('close modal-1'));
  });

  test('restores focus to the last active element', () => {
    const { getByText, rerender } = render(renderSingleModal(false));

    // focus button
    getByText('last focus').focus();

    rerender(renderSingleModal(true));

    vi.runAllTimers();
    expect(document.activeElement).toBe(getByText('close modal-1'));

    rerender(renderSingleModal(false));

    vi.runAllTimers();
    expect(document.activeElement).toBe(getByText('last focus'));
  });

  test('restores focus to the last active element when multiple modals are opened and closed', () => {
    const { getByText, rerender } = render(renderMultiModal(false, false));

    // focus button
    getByText('last focus').focus();

    rerender(renderMultiModal(true, false));

    vi.runAllTimers();
    expect(document.activeElement).toBe(getByText('close modal-1'));

    rerender(renderMultiModal(true, true));

    vi.runAllTimers();
    expect(document.activeElement).toBe(getByText('close modal-2'));

    rerender(renderMultiModal(true, false));

    vi.runAllTimers();
    expect(document.activeElement).toBe(getByText('close modal-1'));

    rerender(renderMultiModal(false, false));

    vi.runAllTimers();
    expect(document.activeElement).toBe(getByText('last focus'));
  });
});
