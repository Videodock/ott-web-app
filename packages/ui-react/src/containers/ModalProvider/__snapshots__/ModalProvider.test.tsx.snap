// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ModalProvider /> > renders provider with test modal 1`] = `
<div>
  <div
    aria-hidden="true"
    data-testid="root"
    id="root"
    inert=""
  >
    <button
      tabindex="-1"
    >
      last focus
    </button>
  </div>
  <div
    data-testid="modal-modal-1"
  >
    <button>
      close 
      modal-1
    </button>
  </div>
</div>
`;
